pkg/dns/service_test.go

```diff
@@ -360,3 +416,497 @@ func TestDNS_findMatchingManagedZone(t *testing.T) {
 		})
 	}
 }
+
+func TestService_CleanupDNSRecords(t *testing.T) {
+
+	tests := []struct {
+		name    string
+		gateway *gatewayv1beta1.Gateway
+		record  *v1alpha1.DNSRecord
+		wantErr bool
+	}{
+		{
+			name: "DNS record gets deleted",
+			gateway: &gatewayv1beta1.Gateway{
+				ObjectMeta: v1.ObjectMeta{
+					UID: types.UID("test"),
+				},
+			},
+			record: &v1alpha1.DNSRecord{
+				ObjectMeta: v1.ObjectMeta{
+					Labels: map[string]string{
+						dns.LabelGatewayReference: "test",
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "no DNS records do be deleted",
+			gateway: &gatewayv1beta1.Gateway{
+				ObjectMeta: v1.ObjectMeta{
+					UID: types.UID("test"),
+				},
+			},
+			record:  &v1alpha1.DNSRecord{},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gw := traffic.NewGateway(tt.gateway)
+			f := fake.NewClientBuilder().WithScheme(testScheme(t)).WithObjects(tt.record).Build()
```

Comment:
not sure if you are aware, but you can use the fake client to get the DNSRecord after deletion and it should be gone. The fake client acts like an in mem store